project/CrossSparkVersions.scala

```diff
@@ -0,0 +1,248 @@
+import Unidoc._
+
+/**
+ * Specification for a Spark version with all its build configuration.
+ *
+ * @param fullVersion The full Spark version (e.g., "3.5.7", "4.0.2-SNAPSHOT")
+ * @param targetJvm Target JVM version (e.g., "11", "17")
+ * @param additionalSourceDir Optional version-specific source directory suffix (e.g., "scala-spark-3.5")
+ * @param antlr4Version ANTLR version to use (e.g., "4.9.3", "4.13.1")
+ * @param additionalJavaOptions Additional JVM options for tests (e.g., Java 17 --add-opens flags)
+ * @param isLatest Whether this is the latest released version (affects artifact naming and doc generation)
+ */
+case class SparkVersionSpec(
+  fullVersion: String,
+  targetJvm: String,
+  additionalSourceDir: Option[String],
+  antlr4Version: String,
+  additionalJavaOptions: Seq[String] = Seq.empty,
+  isLatest: Boolean = false
+) {
+  /** Returns the Spark short version (e.g., "3.5", "4.0") */
+  def shortVersion: String = {
+    Mima.getMajorMinorPatch(fullVersion) match {
+      case (maj, min, _) => s"$maj.$min"
+    }
+  }
+
+  /** Returns log4j config file based on source directory */
+  def log4jConfig: String = {
+    if (additionalSourceDir.exists(_.contains("master"))) "log4j2_spark_master.properties"
+    else "log4j2.properties"
+  }
+
+  /** Whether to export JARs instead of class directories (needed for Spark Connect on master) */
+  def exportJars: Boolean = additionalSourceDir.exists(_.contains("master"))
+
+  /** Whether to generate Javadoc/Scaladoc for this version */
+  def generateDocs: Boolean = isLatest
+}
+
+object SparkVersionSpec {
+  /** Spark 3.5.7 - latest released version */
+  val spark35 = SparkVersionSpec(
+    fullVersion = "3.5.7",
+    targetJvm = "11",
+    additionalSourceDir = Some("scala-spark-3.5"),
+    antlr4Version = "4.9.3",
+    additionalJavaOptions = Seq.empty,
+    isLatest = true
+  )
+
+  /** Spark 4.0.2-SNAPSHOT - master branch */
+  val spark40 = SparkVersionSpec(
+    fullVersion = "4.0.2-SNAPSHOT",
+    targetJvm = "17",
+    additionalSourceDir = Some("scala-spark-master"),
+    antlr4Version = "4.13.1",
+    additionalJavaOptions = Seq(
+      // Copied from SparkBuild.scala to support Java 17 for unit tests (see apache/spark#34153)
+      "--add-opens=java.base/java.lang=ALL-UNNAMED",
+      "--add-opens=java.base/java.lang.invoke=ALL-UNNAMED",
+      "--add-opens=java.base/java.io=ALL-UNNAMED",
+      "--add-opens=java.base/java.net=ALL-UNNAMED",
+      "--add-opens=java.base/java.nio=ALL-UNNAMED",
+      "--add-opens=java.base/java.util=ALL-UNNAMED",
+      "--add-opens=java.base/java.util.concurrent=ALL-UNNAMED",
+      "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED",
+      "--add-opens=java.base/sun.nio.cs=ALL-UNNAMED",
+      "--add-opens=java.base/sun.security.action=ALL-UNNAMED",
+      "--add-opens=java.base/sun.util.calendar=ALL-UNNAMED"
+    ),
+    isLatest = false
+  )
+
+  /** Latest released Spark version */
+  val LATEST_RELEASED = spark35
```

Comment:
It's kind of weird that sometimes we determine if `isLatest` by comparing with this, sometimes via the spec arg `isLatest` seems like two SOT for this. We can probably remove one.

Reply:
yes. good catch. removed.